File: 115-distinct-subsequences/115-distinct-subsequences.java
class Solution {
    public int numDistinct(String s, String t) {
        int n=s.length();
        int m=t.length();
        
        int[] curr = new int[m + 1];
        
        curr[0]=1;
        
        for(int i=1;i<=n;i++)
            for(int j=m;j>=1;j--)
                if(s.charAt(i-1) == t.charAt(j-1)) curr[j] = curr[j-1] + curr[j];
        
        return (int)curr[m];
    }
}

File: 115-distinct-subsequences/115-distinct-subsequences.cpp
class Solution {
public:
    int numDistinct(string s, string t) {
        int n = s.length(), m = t.length();
        
        vector<vector<int>> dp(n + 1, vector<int>(m + 1, 0));
        
        for(int i=0;i<n;i++)
            dp[i][0] = 1;
        
        for(int i=1;i<=n;i++) {
            for(int j=1;j<=m;j++) {
                long long count = 0;
                if(s[i - 1] == t[j - 1]) count += dp[i - 1][j - 1];
                count += dp[i - 1][j];
                dp[i][j] = count;
            }
        }
        return dp[n][m];
    }
};